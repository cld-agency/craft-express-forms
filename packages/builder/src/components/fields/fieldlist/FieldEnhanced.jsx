import PropTypes from 'prop-types';
import React from 'react';
import translate from '../../../functions/translator';
import ToggleSwitch from '../../common/lightswitch/ToggleSwitch';
import FileExtras from './extras/FileExtras';
import RadialActionButton from './RadialActionButton';
import { FILE } from '../types';

const FieldEnhanced = ({
  type,
  name,
  handle,
  required,
  nameHandler,
  handleHandler,
  customFieldPropertyHandler,
  requiredHandler,
  typeHandler,
  removeHandler,
  dragHandleProps,
  ...extraProps
}) => (
  <>
    <div className="base">
      <div className="main-row">
        <div className="icon">
          <RadialActionButton type={type} changeTypeHandler={typeHandler} />
        </div>
        <div className="data">
          <div>
            <input
              className={`name ${name ? 'has-content' : ''}`}
              type="text"
              value={name ?? ''}
              onChange={nameHandler}
              autoComplete="off"
            />
            <label>{translate('Name')}</label>
            <span className="focus-border" />
          </div>
          <div>
            <input
              className={`handle ${handle ? 'has-content' : ''}`}
              type="text"
              value={handle ?? ''}
              onChange={handleHandler}
              autoComplete="off"
            />
            <label>{translate('Handle')}</label>
            <span className="focus-border" />
          </div>
        </div>
        <div className="attributes">
          <div>
            <label>{translate('Required')}</label>
            <div>
              <ToggleSwitch value={required} toggleHandler={requiredHandler} />
            </div>
          </div>
        </div>
      </div>

      {type === FILE && <FileExtras {...extraProps} propChangeHandler={customFieldPropertyHandler} />}
    </div>
    <div className="actions">
      <a
        className="icon delete"
        onClick={() => {
          const confirmMessage = 'Are you sure you want to remove this field?';
          if (confirm(translate(confirmMessage))) {
            removeHandler();
          }
        }}
        tabIndex={-1}
      />
      <a className="icon move" {...dragHandleProps} tabIndex={-1} />
    </div>
  </>
);

FieldEnhanced.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string,
  handle: PropTypes.string,
  required: PropTypes.bool,
  nameHandler: PropTypes.func,
  handleHandler: PropTypes.func,
  requiredHandler: PropTypes.func,
  typeHandler: PropTypes.func,
  removeHandler: PropTypes.func,
  customFieldPropertyHandler: PropTypes.func,
  dragHandleProps: PropTypes.object,
};

export default FieldEnhanced;
