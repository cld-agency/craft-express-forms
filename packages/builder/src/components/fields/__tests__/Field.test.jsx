import { shallow } from 'enzyme';
import React from 'react';
import sinon from 'sinon';
import FieldEnhanced from '../fieldlist/FieldEnhanced';

describe('<Field />', () => {
  it('renders & displays correctly', () => {
    const wrapper = shallow(<FieldEnhanced name="Field" handle="field" type="text" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders & displays correctly with empty name and handle', () => {
    const wrapper = shallow(<FieldEnhanced type="text" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('changes name', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<FieldEnhanced name="Field" handle="field" type="text" nameHandler={onChange} />);

    wrapper.find('input.name').simulate('change', 'test');
    expect(onChange).toBeCalledWith('test');
  });

  it('changes handle', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<FieldEnhanced name="Field" handle="field" type="text" handleHandler={onChange} />);

    wrapper.find('input.handle').simulate('change', 'test');
    expect(onChange).toBeCalledWith('test');
  });

  it('shows extras for File type fields', () => {
    const wrapper = shallow(<FieldEnhanced name="Field" handle="field" type="file" />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Connect(FileExtras)')).toHaveLength(1);
  });

  describe('confirm dialog', () => {
    it('opens a confirm dialog on delete', () => {
      const confirm = sinon.spy();
      global.confirm = confirm;
      const wrapper = shallow(<FieldEnhanced name="Field" handle="field" type="text" />);

      wrapper.find('a.delete').simulate('click');
      expect(confirm.called);
      global.confirm = undefined;
    });

    it('calls the removeHandler if confirmed', () => {
      const confirm = sinon.fake.returns(true);
      const removeHandler = sinon.spy();

      global.confirm = confirm;
      const wrapper = shallow(<FieldEnhanced name="Field" handle="field" type="text" removeHandler={removeHandler} />);

      wrapper.find('a.delete').simulate('click');
      expect(confirm.called).toBeTruthy();
      expect(removeHandler.called).toBeTruthy();
      global.confirm = undefined;
    });

    it('does not call the removeHandler if not confirmed', () => {
      const confirm = sinon.fake.returns(false);
      const removeHandler = sinon.spy();

      global.confirm = confirm;
      const wrapper = shallow(<FieldEnhanced name="Field" handle="field" type="text" removeHandler={removeHandler} />);

      wrapper.find('a.delete').simulate('click');
      expect(confirm.called).toBeTruthy();
      expect(removeHandler.called).toBeFalsy();
      global.confirm = undefined;
    });
  });
});
