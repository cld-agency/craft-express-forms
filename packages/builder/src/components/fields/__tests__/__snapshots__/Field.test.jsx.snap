// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> renders & displays correctly 1`] = `
<Fragment>
  <div
    className="base"
  >
    <div
      className="main-row"
    >
      <div
        className="icon"
      >
        <RadialActionButton
          type="text"
        />
      </div>
      <div
        className="data"
      >
        <div>
          <input
            autoComplete="off"
            className="name has-content"
            type="text"
            value="Field"
          />
          <label>
            Name
          </label>
          <span
            className="focus-border"
          />
        </div>
        <div>
          <input
            autoComplete="off"
            className="handle has-content"
            type="text"
            value="field"
          />
          <label>
            Handle
          </label>
          <span
            className="focus-border"
          />
        </div>
      </div>
      <div
        className="attributes"
      >
        <div>
          <label>
            Required
          </label>
          <div>
            <ToggleSwitch />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="actions"
  >
    <a
      className="icon delete"
      onClick={[Function]}
      tabIndex={-1}
    />
    <a
      className="icon move"
      tabIndex={-1}
    />
  </div>
</Fragment>
`;

exports[`<Field /> renders & displays correctly with empty name and handle 1`] = `
<Fragment>
  <div
    className="base"
  >
    <div
      className="main-row"
    >
      <div
        className="icon"
      >
        <RadialActionButton
          type="text"
        />
      </div>
      <div
        className="data"
      >
        <div>
          <input
            autoComplete="off"
            className="name "
            type="text"
            value=""
          />
          <label>
            Name
          </label>
          <span
            className="focus-border"
          />
        </div>
        <div>
          <input
            autoComplete="off"
            className="handle "
            type="text"
            value=""
          />
          <label>
            Handle
          </label>
          <span
            className="focus-border"
          />
        </div>
      </div>
      <div
        className="attributes"
      >
        <div>
          <label>
            Required
          </label>
          <div>
            <ToggleSwitch />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="actions"
  >
    <a
      className="icon delete"
      onClick={[Function]}
      tabIndex={-1}
    />
    <a
      className="icon move"
      tabIndex={-1}
    />
  </div>
</Fragment>
`;

exports[`<Field /> shows extras for File type fields 1`] = `
<Fragment>
  <div
    className="base"
  >
    <div
      className="main-row"
    >
      <div
        className="icon"
      >
        <RadialActionButton
          type="file"
        />
      </div>
      <div
        className="data"
      >
        <div>
          <input
            autoComplete="off"
            className="name has-content"
            type="text"
            value="Field"
          />
          <label>
            Name
          </label>
          <span
            className="focus-border"
          />
        </div>
        <div>
          <input
            autoComplete="off"
            className="handle has-content"
            type="text"
            value="field"
          />
          <label>
            Handle
          </label>
          <span
            className="focus-border"
          />
        </div>
      </div>
      <div
        className="attributes"
      >
        <div>
          <label>
            Required
          </label>
          <div>
            <ToggleSwitch />
          </div>
        </div>
      </div>
    </div>
    <Connect(FileExtras) />
  </div>
  <div
    className="actions"
  >
    <a
      className="icon delete"
      onClick={[Function]}
      tabIndex={-1}
    />
    <a
      className="icon move"
      tabIndex={-1}
    />
  </div>
</Fragment>
`;
